### ISSUE TITLE
cmd/evm: disasm can't handle valid bytecode with data at the end

### ISSUE THREAD

**Issue by inian** [ethereum/go-ethereum#14376]
#### System information

Geth version: 1.5.9-stable
evm version: 1.6.0-stable-facc47cb
OS & Version: Ubuntu 16.04

`disasm` is unable to handle theDAO bytecode. The bytecode can be found [here](https://github.com/slockit/DAO/wiki/The-DAO-v1.0-Code)

```
evm disasm dao.evm
```

#### Expected behaviour
See the disassembled output

#### Actual behaviour
disasm crashes with this stacktrace - https://pastebin.com/zx48s5ax

Based on the stackoverflow answer [here](https://ethereum.stackexchange.com/a/15289/7116), looks like `disasm` is unable to handle some of the optimizations performed by the solidity compiler

**Comment by fjl** [comment]
This is not a crash (and the error is valid), but I see that this is inconvenient because bytecode with embedded data exists and one should be able to disassemble it.

**Comment by inian** [comment]
You are right, it is not a crash with the latest version of evm. It was a crash in the previous versions which has been handled now. 

But as you mentioned, it would be great to add support for disassembling these kinds of contracts too.

**Comment by loiluu** [comment]
@fjl any update on this issue? Our Oyente project relies heavily on the disassambler, and many people have reported the `missing opcodes` problem.

**Comment by nolash** [comment]
I still get this "crash" with latest version of evm;

https://gist.github.com/nolash/62e6c872ec126594359d20cbe26cd3aa

What is this data at the end? I've also asked here:

https://ethereum.stackexchange.com/questions/23525/what-is-the-cryptic-part-at-the-end-of-a-solidity-contract-bytecode

**Comment by herrBez** [comment]
Any news about this issue?

**Comment by gballet** [comment]
@inian the DAO binary that you speak of not being available anymore, I went to check the [code on etherscan](https://etherscan.io/address/0xbb9bc244d798123fde783fcc1c72d3bb8c189413#code) and indeed the payload gets decoded as if it were part of the main binary stream, and that confuses the disassembler. Can you confirm that this is the right file?

**Comment by inian** [comment]
@gballet yup thats the right file.

**Comment by gballet** [comment]
The thing is, code and data are the same thing. If you know the offset beforehand, you could disassemble from a given offset and get the payload separately.

**Comment by holiman** [comment]
Eh? Code and data are not the same thing

**Comment by holiman** [comment]
I just looked at the original report. This is correct behaviour, from an evm-level perspective the 'junk' data that solidity puts at the end of the constructor is disassembled as code, and eventually hits upon a `PUSHXX` instruction which has too little data to be valid. There's nothing else we can do at this point. In order to do something more 'clever', the disassembler would need more contextual info, like @gballet mentioned, to figure out which offset to start parsing from. 
A user who has that info can already just give the correctly offsetted bytecode to the disassembler. 

Closing this, I don't see this as a bug.
